lib/cloud_text.rb

```diff
@@ -9,47 +9,64 @@ def self.clean_text(input, options = {})
     @input = input
     @options = options
 
-    # Remove punctuation and/or digits
+    # Get feature on/off switches
     remove_digits = @options.fetch(:remove_digits, false)
-    if remove_digits
-      @input = input.gsub(/[^A-Za-z0-9^şŞıİçÇöÖüÜĞğ\s]|_|\d/, ' ')
-    else
-      @input = input.gsub(/[^A-Za-z0-9^şŞıİçÇöÖüÜĞğ\s]|_/, ' ')
-    end
-    # Reduce multiple whitespaces into single whitespace
-    @input = @input.gsub(/\s+/, ' ')
+    stemming_enabled = @options.fetch(:stemming, false)
 
-    # Lowercase all tokens
+    # Get variables
     @language = @options.fetch(:language, "en")
-    @input = UnicodeUtils.downcase(@input, @language.to_sym)
-
-    # Filter stopwords
-    @cutsom_stopwords = @options.fetch(:stopwords, [])
-    if @language == "tr"
-      stopword_filter = Stopwords::Filter.new tr_stopwords
-    else
-      stopword_filter = Stopwords::Snowball::Filter.new @language
-    end
+    @custom_stopwords = @options.fetch(:stopwords, [])
+
+    @input = process_text(@input, stemming_enabled, remove_digits, @language, @custom_stopwords)
+    count_words(@input)
+  end
+
+  private
+
+  def self.process_text(input, stemming_enabled, remove_digits, language, custom_stopwords)
+    input = remove_punctuation(input, remove_digits)
+    input = reduce_whitespaces(input)
+    input = lowercase_words(input, language)
+    input = filter_stopwords(input, language, custom_stopwords) # Get custom_stopwords from user and filter words
+    input = stemming(input, language) if stemming_enabled
+  end
```

Comment:
only thing that bugs me is the repetition here, we could use method chaining.

Reply:
Can you give me a hand on this? I am not sure how to do this correctly.

Reply:
Absolutely.

Reply:
I just realized we can't do this without monkey-patching some classes. Since input always the first argument, we could check elixir like pipe operator solutions for ruby, but that would mean an another dependency. So, for now, I think this should stay as is.

Reply:
I agree, we'll do it later than?

Reply:
Yeap 👍